sdk/azcore/internal/shared/response_error.go

```diff
@@ -0,0 +1,138 @@
+//go:build go1.16
+// +build go1.16
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package shared
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"regexp"
+)
+
+// NewResponseError creates a new *ResponseError from the provided HTTP response.
+func NewResponseError(resp *http.Response) error {
+	respErr := &ResponseError{
+		StatusCode:  resp.StatusCode,
+		RawResponse: resp,
+	}
+
+	// prefer the error code in the response header
+	if ec := resp.Header.Get("x-ms-error-code"); ec != "" {
+		respErr.ErrorCode = ec
+	}
+
+	// write the request method and URL with response status code
+	msg := &bytes.Buffer{}
```

Comment:
I think it would be better to do all this work in the Error method because you ALWAYS do it when an error occurs here but only do it if Error() is called by the customer. This code creates more heap objects and takes CPU time that is really only needed if Error() is called by the customer's code.

Reply:
While I agree, the case that makes this tricky is when the error code is in the body, not the `x-ms-error-code` header.  We don't want `Error()` to populate the `ErrorCode` post-construction.  So if we have to do this work in the constructor, we end up doing almost the same work in `Error()` as well (more garbage etc).

This keeps things simple, and I think it's highly likely that `Error()` will be invoked.  If this really becomes a problem, we can change it later.